### ISSUE TITLE
Selectable Quirks

### ISSUE THREAD

**Issue by mansellan** [WaynePhillipsEA/twinbasic#215]
**Is your feature request related to a problem?**
As seen in #209, some of the "Quirks" in VBx are probably more accurately called "bugs". Whilst it's important (for backwards compatibility) to replicate these as closely as possible, I wonder if there's value in being able to selectively disable quirk-adherence for a less buggy experience.

**Describe the solution you'd like**
Ability to be able to selectively disable quirk-adherence for each quirky area. For example, a setting for Like, a setting for... ? This probably doesn't need to be any more granular than a project-level setting.

**Describe alternatives you've considered**
Living with and working around the quirks from VBx which have been faithfully re-implemented in tB.

**Additional context**
Not sure if all the known quirks are listed anywhere?

**Comment by WaynePhillipsEA** [comment]
Thanks for opening this issue.  Yes, this is planned, and I suggest all quirks will want to be turned off in new projects.  The Like operator is most definitely a prime candidate for this.

**Comment by Kr00l** [comment]
My proposal (2 cent) for that "quirk mode" setting would be a "Flags" value. (OR'ed Long's)
These "Flags" are saved in one setting and can be represented in a "reference" style list to check/uncheck the different quirks.
So it's tidied up and can be extended easily.

_(Deleted this post on #209 and moved to here)_

**Comment by Greedquest** [comment]
What might be a use case for enabling some quirks and not others? That sounds like a recipe for code you can't pass onto another developer very easily (or future you) - code which does one thing in one setting but not another - what about shared libraries and modules - each with a different selection of quirk settings. 

Ok I appreciate you could have a codebase that relies on a single quirk and no others and consequently only checks one box, but IDK how likely that is - and how many people will be able to identify the single quirk their codebase relies on - it's either small enough to identify the few instances and update them when migrating to tB or so big it probably has more quirks in it; this feels like something that should be a single checkbox IMO, or just a feature of "legacy mode".

---

Changing the behaviour of `Like` is a breaking change, are there other examples of breaking changes between legacy mode and current - as in where valid VB* legacy mode code won't do the same thing in tB?

**Comment by bclothier** [comment]
I have been thinking about this and this has been mentioned twice in 2 separate issues...

This feels like Internet Explorer all again. Do you remember how IE7-9 had "compatibility view"? Quick, how many times did any of you go "oh, this isn't working, let me switch on compatibility view" (or vice versa)? I'd wager seldom, if ever. Most of us just shrugged and fired up our Firefox, Chrome, Safari, whatever. Why? It Just Works™. 

tB should just work™ for the same reason. 

For that reason, I'm opposed to the idea of a quirk mode. Because backward compatibility is a priority, we should strive for that but the best solution for addressing quirks is to provide a superior alternative. 

For example, the quirks identified #279 can be best addressed by providing a superior filesystem API that aren't some weird language syntax. Likewise, #469 can be remedied by providing a superior inputbox that also allows for testable code. 

I really really think quirk mode is the expressway to the death by thousands cuts. Please don't go there. Providing a superior alternative provides more benefits:
1) any existing codebase will continue to work as before, with or without any pre-existing workarounds for the quirks of the original feature. Thus it remains familiar to those accustomed to the quirks.
2) superior alternatives provides opportunity to learn the best practice and avoid the anti-patterns inherent in the original feature. 
3) the experience is much simpler as there are no switches to fiddle with, and bug reporting & troubleshooting likewise are simplified on the account of not having to ask "are you running so and so with quirk mode on or off"

I think tB can do without quirk modes.

**Comment by Greedquest** [comment]
@bclothier The risk is that if you have more than one way of doing something in the language, alternate APIs, people won't know which to pick. See [Zen of Python](https://www.python.org/dev/peps/pep-0020/) :

> There should be one-- and preferably only one --obvious way to do it.

That's why it's nice to offer these API improvements as "overloads" of existing language features since then there is only one place to look.

This is just to highlight that point, I'm not sure which is the lesser of the two evils.

**Comment by Kr00l** [comment]
I'm also afraid for the file language syntax in tB.

Will it be unicode but buggy? So, unless you use the new fancy file system objects it's not good?

So somebody migrates a project, thinking all is good now, but not aware of to change all the code base to the new objects..
So I'm thinking it should be all bug free by default and preferable only one way of doing. If there would be two ways possible, like a dialog class, there should be both bug free just the new way offering more possibilities..

But there should not be a scenario:
"The file system syntax has bugs, don't use it".
For VB6 we are used for replacements, often just to have unicode. But that's not a thing anymore in tB.

**Comment by bclothier** [comment]
@Greedquest 

> The risk is that if you have more than one way of doing something in the language, alternate APIs, people won't know which to pick. 

I absolutely agree with the principle of having only one way of doing something (I've argued for that in other places as a matter of fact). However, Python came about as an alternate to the existing scripting languages and therefore backward compatibility was not a design concern for Python. Thus, they had the freedom to do it a pythonic way.

In contrast, tB's design goal is to be backward compatible precisely in order to allow easy porting of VBx codebase into tB. This puts tB in the same position as Microsoft was when they had a chance to fix [the infamous 1900 bug](https://docs.microsoft.com/en-us/office/troubleshoot/excel/wrongly-assumes-1900-is-leap-year) from Lotus123 in Excel. In the end, they decided to _not_ fix the bug simply so that people could use their Lotus123 spreadsheets in Excel. Thus, we need to account for the possibility that fixing bad behaviors will actually harm the backward compatibility, especially if the original codebase already had its own workaround for the bad behavior and now will not function under the changed assumption. 

The other problem is that the original design of VBx is actually an anti-pattern and can hinder any future enhancements. Take file I/Os. For whatever bizarre reasons Microsoft had, they decided to make file I/Os _statements_, not an API. I am not aware of any other languages that makes file I/O a part of the language syntax, but VBx does. That alone brings so many problem -- quick, how do you extend a file open? Oh wait, you can't overload nor shadow it. Do you want Wayne to be the lone maintainer for the file I/O, simply because it's tied up in the compiler & language? I'd rather not. The superior alternative then can be provided via a system similar to nuget (see #16 ), and for special projects, it can be easily extended by overriding or shadowing. 

The other scenario with InputBox is not as a problematic because you can actually shadow it with your own custom InputBox:
```
Public Function InputBox( _
    Prompt As Variant, _
    Optional Title As Variant, _
    Optional Default As Variant, _
    Optional XPos As Variant, _
    Optional YPos As Variant, _
    Optional HelpFile As Variant, _
    Optional Context As Variant _
) As String
    Debug.Print "shadowing Inputbox"
    InputBox = VBA.InputBox(Prompt, Title, Default, XPos, YPos, HelpFile, Context)
End Function
```

Thus, it's easy to "fix" the shortcomings even in VBx. However, when you shadow anything from VBA API, you are still changing how things work and that can break the codebase not expecting anything other than a `VBA.InputBox`. If tB provides a way to replace or extend the message or input boxes, it would be no different than being able to shadow the same in VBA. However, a fundamental difference would be that in VBx, it was common to [hack the Inputbox](https://www.vbforums.com/showthread.php?780945-***-mask-password-in-inputbox) (also see #431 ), which means breaking the encapsulation of the InputBox. Code like those are actually _more_ fragile because they rely on InputBox being a certain way, finding windows and controls by identifiers. That is bad for extensibility because any custom inputbox would then break those code. In those cases, the correct answer ought to be "don't hack like this". So for the sake of backward compatibility, it's probably best to treat `InputBox` as forever frozen in time to support codebases that contains hacking codes to manipulate the Inputbox in an uncontrolled manner, so that it can port to tB and still work. 

But, in new projects (or modules), it would be bad to use them the same way it was used in VBx since we would want to write testable code and avoid mixing UI concerns with logic as VBx code is prone to do. For that reason, we'll need to take a different approach nonetheless. So we end up saying, "don't use that buggy approach." anyway.

**Comment by mansellan** [comment]
I'm kinda torn on this one. I guess my original idea was that no quirks would be enabled by default, the "out-of-the-box" experience would be that a bug-free implementation would be used, but the original buggy behaviour could be explicitly opted into. But I definitely agree that its not optimal, and certainly not how a purely greenfield language would operate.

I do like the idea of emitting compiler warnings though. With that, users can choose to downgrade them to "don't show" if they definitely want the bug, or upgrade them to "compile error" if they never want to accidentally use the affected method.

So I guess on balance, if its possible to solve by providing superior alternatives, that's probably the best route. Kinda like marking the quirky route as [Obsolete] in C#

**Comment by bclothier** [comment]
Yes, and using `Obsolete` can be used to either remove or de-prioritize the intellisense. So if I type in `Open`, the intellisense could show a `File.Open()`, which would then provide me with a hint that I should use this instead of the legacy `Open` statement. I wouldn't be prevented from using the legacy `Open` statement, but I'd be now swimming upstream, shouting "yes, I really really want to use the old buggy quirky syntax!" at the top of my lung. 😄

**Comment by mwolfe02** [comment]
> This puts tB in the same position as Microsoft was when they had a chance to fix [the infamous 1900 bug](https://docs.microsoft.com/en-us/office/troubleshoot/excel/wrongly-assumes-1900-is-leap-year) from Lotus123 in Excel. 

Sorry for the off-topic comment, but I had to post this [Joel Spolsky anecdote](https://www.joelonsoftware.com/2006/06/16/my-first-billg-review/) about that 1900 bug for anyone who hasn't read it yet.
